src/lib.rs

```diff
@@ -78,14 +235,20 @@ impl Header {
 #[cfg(feature = "gecko-ffi")]
 impl Header {
     fn cap(&self) -> usize {
-        (self._cap & CAP_MASK) as usize
+        ((self._cap & CAP_MASK) >> CAP_SHIFT) as usize
     }
 
     fn set_cap(&mut self, cap: usize) {
         debug_assert!(cap & (CAP_MASK as usize) == cap);
         // FIXME: this is busted because it reads uninit memory
         // debug_assert!(!self.uses_stack_allocated_buffer());
-        self._cap = assert_size(cap) & CAP_MASK;
+
+        // NOTE: it's fine that we unconditionally mask out the
+        // mIsAutoArray bit, because the Rust side will never
+        // construct an auto array and set_cap is only invoked
+        // by Rust. Auto arrays only happen when a C++-constructed
+        // nsTArray bridges into a ThinVec.
+        self._cap = (assert_size(cap) << CAP_SHIFT) & CAP_MASK;
```

Comment:
`Header::cap` right shifts by `CAP_SHIFT`; here, we left shift by `CAP_SHIFT`.  This code also inverts the order of masking vs. shifting relative to `Header::cap`.  Why the difference?

Reply:
Loading vs Storing a packed value requires inverted operations, no? (Also note the shift is a no-op on little-endian, and the mask is unnecessary when the shift *isn't* a no-op, because left-shift inserts 0's)